core/src/main/java/jenkins/util/VirtualFile.java

```diff
@@ -334,12 +337,60 @@ private static void collectFiles(VirtualFile d, Collection<String> names, String
+     *                 the empty string means <em>no matches</em> (use {@link SelectorUtils#DEEP_TREE_MATCH} if you want to match everything except some excludes)
+     * @param excludes optional excludes in similar format to {@code includes}
+     * @param useDefaultExcludes as per {@link AbstractFileSet#setDefaultexcludes}
+     * @param noFollowLinks if true then do not follow links.
+     * @param prefix the partial path that will be added before each entry inside the archive.
+     *               If non-empty, a trailing slash will be enforced.
+     * @return the number of files inside the archive (not the folders)
+     * @throws IOException if this is not a directory, or listing was not possible for some other reason
      * @since TODO
      */
-    @Restricted(NoExternalUse.class)
     public int zip(OutputStream outputStream, String includes, String excludes, boolean useDefaultExcludes,
-                            boolean noFollowLinks, String prefix) throws IOException, UnsupportedOperationException {
-        throw new UnsupportedOperationException("Not implemented.");
+                   boolean noFollowLinks, String prefix) throws IOException {
+        String correctPrefix;
+        if (StringUtils.isBlank(prefix)) {
+            correctPrefix = "";
+        } else {
+            correctPrefix = Util.ensureEndsWith(prefix, "/");
+        }
+        
+        Collection<String> files = list(includes, excludes, useDefaultExcludes, noFollowLinks);
+        try (ZipOutputStream zos = new ZipOutputStream(outputStream)) {
+            zos.setEncoding(System.getProperty("file.encoding")); // TODO JENKINS-20663 make this overridable via query parameter
+
+            for (String relativePath : files) {
+                VirtualFile virtualFile = this.child(relativePath);
+                sendOneZipEntry(zos, virtualFile, relativePath, noFollowLinks, correctPrefix);
+            }
+        }
+        return files.size();
+    }
+
+    private void sendOneZipEntry(ZipOutputStream zos, VirtualFile vf, String relativePath, boolean noFollowLinks, String prefix) throws IOException {
+        // In ZIP archives "All slashes MUST be forward slashes" (http://pkware.com/documents/casestudies/APPNOTE.TXT)
+        // TODO On Linux file names can contain backslashes which should not treated as file separators.
+        //      Unfortunately, only the file separator char of the master is known (File.separatorChar)
+        //      but not the file separator char of the (maybe remote) "dir".
+        String onlyForwardRelativePath = relativePath.replace('\\', '/');
+        String zipEntryName = prefix + onlyForwardRelativePath;
+        ZipEntry e = new ZipEntry(zipEntryName);
+
+        e.setTime(vf.lastModified());
+        zos.putNextEntry(e);
+        try (InputStream in = vf.open(noFollowLinks)) {
+            // hudson.util.IOUtils is already present
```

Comment:
Huh?

Reply:
Explaining the un-imported call with fully specified class name below?

Reply:
OK, “present” did not click with me.